pdfiumandroid/src/main/java/io/legere/pdfiumandroid/util/PdfiumNativeSourceBridge.kt

```diff
@@ -0,0 +1,28 @@
+package io.legere.pdfiumandroid.util
+
+import io.legere.pdfiumandroid.BuildConfig
+import io.legere.pdfiumandroid.PdfiumSource
+
+internal class PdfiumNativeSourceBridge(private val source: PdfiumSource) {
+
+    private var buffer: ByteArray? = null
+
+    fun read(position: Long, size: Long): Int =
+        try {
+            require(size <= Int.MAX_VALUE) { "size is too large" }
+            val trimmedSize = size.toInt()
+
+            var buffer = buffer
+            if (buffer == null || buffer.size < size) {
+                buffer = ByteArray(trimmedSize).also { this.buffer = it }
+            }
+
+            val bytesRead = source.read(position, buffer, trimmedSize)
+            if (bytesRead <= 0) 0 else bytesRead // Pdfium expects 0 for error while Java/Kotlin usually return a negative value
```

Comment:
I decided to stick with Kotlin idioms for Source API. Most developers expect the read() to return negative value if there's error, and they could make a mistake if we make API different.

Reply:
This brings back arrow.  The idea with arrow is don't throw exceptions, and don't use special values for errors.  return an Either<Error, Value>. so if it has a value, it's a good value.  I've become a big fan.